text/0000-bee-network/0000-bee-network.md

```diff
@@ -0,0 +1,295 @@
+
+## Message passing
+
+This crate favors message passing via channels over globally shared state and locks. Instead of keeping the list of endpoints in a globally accessible hashmap this crate separates and moves such state into workers that run asynchronously, and are listening for commands and events to act upon, and also notify listeners by sending events.
+
+## Initialization
+
+In order for this crate to be used, it has to be initialized first by providing a `NetworkConfig`:
+
+```rust
+struct NetworkConfig {
+    // The server or binding TCP port that remote endpoints can connect and send to.
+    binding_port: u16,
+    // The binding address that remote endpoints can connect and send to.
+    binding_addr: IpAddr,
+    // The interval between two connections attempts in case of connection losses.
+    reconnect_interval: u64,
+    // The maximum message length in terms of bytes.
+    max_message_length: usize
+
+    ...
+}
+```
+
+This has the advantage of being easily extensible and keeping the signature of the static `init` function small:
+
+```rust
+fn init(config: NetworkConfig) -> (Network, Events, Shutdown) { ... }
+```
+
+This function returns a tuple struct that allows the consumer of this crate to send commands (`Network`), listen to events (`Events`), and gracefully shutdown all running asynchronous tasks that were spawned within this system (`Shutdown`).
+
+## `Port`, `Address`, `Protocol`, and `Url`
+
+In order to send a message to an endpoint, a node needs to know the endpoint's address and the communication protocol it uses. This crate provides the following abstractions to deal with that:
+
+* `Port`: a 0-cost wrapper around a `u16`, which is introduced for type safety and better readability:
+
+    ```rust
+    struct Port(pub u16);
+    ```
+
+    For convenience, `Port` dereferences to `u16`.
+
+* `Address`: a 0-cost wrapper around a `SocketAddr` which provides an adjusted, but overall simpler API than its inner type:
+
+    ```rust
+    #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)
+    struct Address(SocketAddr);
+    ```
+
+    An `Address` can be constructed from Ipv4 and Ipv6 addresses and a port:
+
+    ```rust
+    // Infallible construction of an `Address` from Ipv4.
+    fn from_v4_addr_and_port(address: Ipv4Addr, port: Port) -> Self { ... }
+
+    // Infallible construction of an `Address` from Ipv6.
+    fn from_v6_addr_and_port(address: Ipv6Addr, port: Port) -> Self { ... }
+
+    // Fallible construction of an `Address` from `&str`.
+    async fn from_addr_str(address: &str) -> Result<Self> { ... }
+    ```
+
+    Note, that the last function is `async`. That is because of possible delays when performing domain name resolution.
+
+* `Protocol`: an `enum`eration of supported communication protocols:
+
+    ```rust
+    #[non_exhaustive]
+    pub enum Protocol {
```

Comment:
This `enum` is `pub`, the next one is not... ?

Reply:
Ooooh, you got me there! :trollface:

Reply:
Removed.